cpp/src/arrow/record_batch.h

```diff
@@ -207,6 +208,14 @@ class ARROW_EXPORT RecordBatchReader {
   /// \return Status
   virtual Status ReadNext(std::shared_ptr<RecordBatch>* batch) = 0;
 
+  // Fallback to sync implementation until all other readers are converted(ARROW-11770)
+  // and then this could become pure virtual with ReadNext falling back to async impl.
+  virtual Future<std::shared_ptr<RecordBatch>> ReadNextAsync() {
+    std::shared_ptr<RecordBatch> batch;
+    ARROW_RETURN_NOT_OK(ReadNext(&batch));
+    return Future<std::shared_ptr<RecordBatch>>::MakeFinished(std::move(batch));
+  }
+
```

Comment:
It looks like across CSV, Parquet, and Feather, we now have two distinct approaches to async reading: here we add a method to asynchronously read the next batch, while in Parquet/Feather we add a method to convert a reader to a generator of batches. We should probably pick one for consistency's sake.

Reply:
It was already different.  The streaming CSV reader has a `ReadBatch` method and it was used to create an iterator of batches.  I think keeping the generator logic out of the readers would be ideal.